README.md
__NOTE__: This is a fork of the original https://github.com/linkedin/URL-Detector, which seems to be abandoned.  This fork was created to allow continued maintenance by the OS community.

# URL Detector

The url detector is a library created by the Linkedin Security Team to detect and extract urls in a long piece of text.

It is able to find and detect any urls such as:

* __HTML 5 Scheme__   - //www.linkedin.com
* __Usernames__       - user:pass@linkedin.com
* __Email__           - fred@linkedin.com
* __IPv4 Address__    - 192.168.1.1/hello.html
* __IPv4 Octets__     - 0x00.0x00.0x00.0x00
* __IPv4 Decimal__    - http://123123123123/
* __IPv6 Address__    - ftp://[::]/hello
* __IPv4-mapped IPv6 Address__  - http://[fe30:4:3:0:192.3.2.1]/

_Note: Keep in mind that for security purposes, its better to overdetect urls and check more against blacklists than to not detect a url that was submitted. As such, some things that we detect might not be urls but somewhat look like urls. Also, instead of complying with RFC 3986 (http://www.ietf.org/rfc/rfc3986.txt), we try to detect based on browser behavior, optimizing detection for urls that are visitable through the address bar of Chrome, Firefox, Internet Explorer, and Safari._

It is also able to identify the parts of the identified urls. For example, for the url: `http://user@linkedin.com:39000/hello?boo=ff#frag`

* Scheme   - "http"
* Username - "user"
* Password - null
* Host     - "linkedin.com"
* Port     - 39000
* Path     - "/hello"
* Query    - "?boo=ff"
* Fragment - "#frag"

---
## How to Use:

Using the URL detector library is simple. Simply import the UrlDetector object and give it some options. In response, you will get a list of urls which were detected.

For example, the following code will find the url `linkedin.com`

```java

    UrlDetector parser = new UrlDetector("hello this is a url Linkedin.com", UrlDetectorOptions.Default);
    List<Url> found = parser.detect();

    for(Url url : found) {
        System.out.println("Scheme: " + url.getScheme());
        System.out.println("Host: " + url.getHost());
        System.out.println("Path: " + url.getPath());
    }
```

### Quote Matching and HTML
Depending on your input string, you may want to handle certain characters in a special way. For example if you are
parsing HTML, you probably want to break out of things like quotes and brackets. For example, if your input looks like

> &lt;a href="http://linkedin.com/abc"&gt;linkedin.com&lt;/a&gt;

You probably want to make sure that the quotes and brackets are extracted. For that reason, using UrlDetectorOptions
will allow you to change the sensitivity level of detection based on your expected input type. This way you can detect
`linkedin.com` instead of `linkedin.com</a>`.

In code this looks like:

```java

    UrlDetector parser = new UrlDetector("<a href="linkedin.com/abc">linkedin.com</a>", UrlDetectorOptions.HTML);
    List<Url> found = parser.detect();

```

### Maven Usage:

To use the latest release, add the following dependency to your pom.xml:

```xml
    <dependency>
        <groupId>io.github.url-detector</groupId>
        <artifactId>url-detector</artifactId>
        <version>1.0.0-SNAPSHOT</version>
    </dependency>
```

---
## About:

This library was written by the security team and Linkedin when other options did not exist. Some of the primary authors are:

* Vlad Shlosberg (vshlosbe@linkedin.com)
* Tzu-Han Jan (tjan@linkedin.com)
* Yulia Astakhova (jastakho@linkedin.com)

---
## Third Party Dependencies

### TestNG
* http://testng.org/
* Copyright © 2004-2014 Cédric Beust
* License: Apache 2.0

### Apache CommonsLang3: org.apache.commons:commons-lang3:3.1
* http://commons.apache.org/proper/commons-lang/
* Copyright © 2001-2014 The Apache Software Foundation
* License: Apache 2.0

---
## License

Copyright 2015 LinkedIn Corp. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the license at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.



CHANGELOG.md
v1.0.0
------

* Fixed: https://github.com/linkedin/URL-Detector/pull/31
* Fixed: https://github.com/URL-Detector/URL-Detector/issues/10

v0.1.23
-------

* Fixed: https://github.com/URL-Detector/URL-Detector/issues/6

v0.1.22
-------

* Fixed: https://github.com/URL-Detector/URL-Detector/issues/5

v0.1.21
-------

* Fixed: https://github.com/URL-Detector/URL-Detector/issues/2

v0.1.20
-------

* Fixed: https://github.com/linkedin/URL-Detector/issues/15

v0.1.19
-------

* Fixed: https://github.com/linkedin/URL-Detector/issues/13

v0.1.18
-------

* Fixed: https://github.com/linkedin/URL-Detector/issues/12
* Fixed: https://github.com/linkedin/URL-Detector/issues/16
* Fixed: https://github.com/linkedin/URL-Detector/issues/19
* Fixed: https://github.com/linkedin/URL-Detector/issues/24

v0.1.17
-------

* Added pom.xml, uploaded to maven central


url-detector/src/test/java/com/linkedin/urls/TestNormalizedUrl.java
/**
 * Copyright 2015 LinkedIn Corp. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 */
package com.linkedin.urls;

import java.net.MalformedURLException;
import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;


public class TestNormalizedUrl {

  @DataProvider
  private Object[][] getHostPathUrls() {
    return new Object[][] {
        {"http://www.google.com/", "www.google.com", "/"},
        {"teeee.com", "teeee.com", "/"},
        {"[::1]", "[0:0:0:0:0:0:0:1]", "/"},
        {"yahoo.com/@1234", "yahoo.com", "/@1234"},
        {"http://[::0xfe.07.23.33]/%25%32%35", "[0:0:0:0:0:0:fe07:1721]", "/%25"},
        {"http://host.com/%2525252525252525", "host.com", "/%25"},
        {"http://[::1]/asdf%25%32%35asd", "[0:0:0:0:0:0:0:1]", "/asdf%25asd"},
        {"http://[::10]/%%%25%32%35asd%%", "[0:0:0:0:0:0:0:10]", "/%25%25%25asd%25%25"},
        {"343324381/", "20.118.182.221", "/"}
    };
  }

  @Test(dataProvider = "getHostPathUrls")
  public void testUsernamePasswordUrls(String testString, String host, String path) throws MalformedURLException {
    Url url = NormalizedUrl.create(testString);
    Assert.assertNotNull(url);
    Assert.assertEquals(url.getHost(), host);
    Assert.assertEquals(url.getPath(), path);
  }

  /**
   * https://developers.google.com/safe-browsing/developers_guide_v3#Canonicalization
   * @return
   */
  @DataProvider
  private Object[][] getFullUrls() {
    return new Object[][] {
        {"http://host/%25%32%35", "http://host/%25"},
        {"http://host/%25%32%35%25%32%35", "http://host/%25%25"},
        {"http://host/%2525252525252525", "http://host/%25"},
        {"http://host/asdf%25%32%35asd", "http://host/asdf%25asd"},
        {"http://host/%%%25%32%35asd%%", "http://host/%25%25%25asd%25%25"},
        {"http://www.google.com/", "http://www.google.com/"},
        {"http://%31%36%38%2e%31%38%38%2e%39%39%2e%32%36/%2E%73%65%63%75%72%65/%77%77%77%2E%65%62%61%79%2E%63%6F%6D/",
            "http://168.188.99.26/.secure/www.ebay.com/"},
        {"http://195.127.0.11/uploads/%20%20%20%20/.verify/.eBaysecure=updateuserdataxplimnbqmn-xplmvalidateinfoswqpcmlx=hgplmcx/",
            "http://195.127.0.11/uploads/%20%20%20%20/.verify/.eBaysecure=updateuserdataxplimnbqmn-xplmvalidateinfoswqpcmlx=hgplmcx/"},
        {"http://host%23.com/%257Ea%2521b%2540c%2523d%2524e%25f%255E00%252611%252A22%252833%252944_55%252B",
            "http://host%23.com/~a!b@c%23d$e%25f^00&11*22(33)44_55+"},
        {"http://3279880203/blah", "http://195.127.0.11/blah"},
        {"http://www.google.com/blah/..", "http://www.google.com/"},
        {"www.google.com/", "http://www.google.com/"},
        {"www.google.com", "http://www.google.com/"},
        {"http://www.evil.com/blah#frag", "http://www.evil.com/blah"},
        {"http://www.GOOgle.com/", "http://www.google.com/"},
        {"http://www.google.com/foo\tbar\rbaz\n2", "http://www.google.com/foobarbaz2"},
        {"http://www.google.com/q?", "http://www.google.com/q?"},
        {"http://www.google.com/q?r?", "http://www.google.com/q?r?"},
        {"http://www.google.com/q?r?s", "http://www.google.com/q?r?s"},
        {"http://evil.com/foo#bar#baz", "http://evil.com/foo"},
        {"http://evil.com/foo;", "http://evil.com/foo;"},
        {"http://evil.com/foo?bar;", "http://evil.com/foo?bar;"},
        {"http://\\x01\\x80.com/", "http://%01%80.com/"},
        {"http://notrailingslash.com", "http://notrailingslash.com/"},
        {"http://www.gotaport.com:1234/", "http://www.gotaport.com:1234/"},
        {"  http://www.google.com/  ", "http://www.google.com/"},
        {"http:// leadingspace.com/", "http://%20leadingspace.com/"},
        {"http://%20leadingspace.com/", "http://%20leadingspace.com/"},
        {"%20leadingspace.com/", "http://%20leadingspace.com/"},
        {"https://www.securesite.com/", "https://www.securesite.com/"},
        {"http://host.com/ab%23cd", "http://host.com/ab%23cd"},
        {"http://host.com//twoslashes?more//slashes", "http://host.com/twoslashes?more//slashes"},
        {"http://go.co/a/b/../c", "http://go.co/a/c"},
        {"http://big.big.boss@0xb02067cd/", "http://big.big.boss@176.32.103.205/"},
        {"http://www.0xb02067cd/", "http://www.0xb02067cd/"},
        {"http://012.0xb02067cd/", "http://012.0xb02067cd/"},
        {"http://☀♁♧.com", "http://xn--k3h6hoe.com/"},
        {"http://💄💃💏.ga", "http://xn--ir8hb8a.ga/"},
        {"http://goȪgle.ga", "http://xn--gogle-jdc.ga/"}
    };
  }

  @Test(dataProvider = "getFullUrls")
  public void testFullUrls(String host, String expectedHost) throws MalformedURLException {
    NormalizedUrl normalizedUrl = NormalizedUrl.create(host);
    Assert.assertEquals(normalizedUrl.getFullUrlWithoutFragment(), expectedHost);
  }
}


url-detector/src/test/java/com/linkedin/urls/TestHostNormalizer.java
/**
 * Copyright 2015 LinkedIn Corp. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 */
package com.linkedin.urls;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Arrays;
import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;


public class TestHostNormalizer {

  @DataProvider
  private Object[][] getIPAddresses() {
    return new Object[][] {
        {"[fefe::]", "[fefe:0:0:0:0:0:0:0]"},
        {"[::ffff]", "[0:0:0:0:0:0:0:ffff]"},
        {"[::255.255.255.255]", "[0:0:0:0:0:0:ffff:ffff]"},
        {"[::]", "[0:0:0:0:0:0:0:0]"},
        {"[::1]", "[0:0:0:0:0:0:0:1]"},
        {"[aAaA::56.7.7.5]", "[aaaa:0:0:0:0:0:3807:705]"},
        {"[BBBB:ab:f78F:f:DDDD:bab:56.7.7.5]", "[bbbb:ab:f78f:f:dddd:bab:3807:705]"},
        {"[Aaaa::1]", "[aaaa:0:0:0:0:0:0:1]"},
        {"[::192.167.2.2]", "[0:0:0:0:0:0:c0a7:202]"},
        {"[0:ffff::077.0x22.222.11]", "[0:ffff:0:0:0:0:3f22:de0b]"},
        {"[0::ffff:077.0x22.222.11]", "63.34.222.11"},
        {"192.168.1.1", "192.168.1.1"},
        {"0x92.168.1.1", "146.168.1.1"},
        {"3279880203", "195.127.0.11"}
    };
  }

  @Test(dataProvider = "getIPAddresses")
  public void testIpHostNormalizationAndGetBytes(String original, String expectedHost) throws UnknownHostException {
    HostNormalizer hostNormalizer = new HostNormalizer(original);
    Assert.assertEquals(hostNormalizer.getNormalizedHost(), expectedHost);

    InetAddress address = InetAddress.getByName(expectedHost);
    byte[] expectedBytes;
    if (address instanceof Inet4Address) {
      expectedBytes = new byte[16];
      expectedBytes[10] = (byte) 0xff;
      expectedBytes[11] = (byte) 0xff;
      System.arraycopy(address.getAddress(), 0, expectedBytes, 12, 4);
    } else {
      expectedBytes = address.getAddress();
    }
    Assert.assertTrue(Arrays.equals(hostNormalizer.getBytes(), expectedBytes));
  }

  @DataProvider
  private Object[][] getNormalHosts() {
    return new Object[][] {
        {"sALes.com"},
        {"33r.nEt"},
        {"173839.com"},
        {"192.168.-3.1"},
        {"[::-34:50]"},
        {"[-34::192.168.34.-3]"}
    };
  }

  @Test(dataProvider = "getNormalHosts")
  public void testSanityAddresses(String host) {
    HostNormalizer hostNormalizer = new HostNormalizer(host);
    Assert.assertEquals(hostNormalizer.getNormalizedHost(), host.toLowerCase());
    Assert.assertNull(hostNormalizer.getBytes());
  }

  @DataProvider
  private Object[][] getUnNormalHosts() {
    return new Object[][] {
      {"☀♁♧.com", "xn--k3h6hoe.com"},
      {"🐩🤠🖊🍩🏳🐯🕶🐋.🍕💩.ws", "xn--gj8huimcrf6a45n5dx91a.xn--vi8hiv.ws"},
      {"goȪgle.ga", "xn--gogle-jdc.ga"}
    };
  }

  @Test(dataProvider = "getUnNormalHosts")
  public void testHostnameNormalization(String unNormalizedHost, String expectedNormalizedHost) {
    HostNormalizer hostNormalizer = new HostNormalizer(unNormalizedHost);
    Assert.assertEquals(hostNormalizer.getNormalizedHost(), expectedNormalizedHost);
  }
}


url-detector/src/main/java/com/linkedin/urls/UrlMarker.java
/**
 * Copyright 2015 LinkedIn Corp. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 */
package com.linkedin.urls;

public class UrlMarker {

  private String _originalUrl;
  private int _absoluteIndex = -1;
  private int _schemeIndex = -1;
  private int _usernamePasswordIndex = -1;
  private int _hostIndex = -1;
  private int _portIndex = -1;
  private int _pathIndex = -1;
  private int _queryIndex = -1;
  private int _fragmentIndex = -1;

  public UrlMarker() {
  }

  public Url createUrl() {
    return new Url(this);
  }

  public void setOriginalUrl(String originalUrl) {
    _originalUrl = originalUrl;
  }

  public String getOriginalUrl() {
    return _originalUrl;
  }

  public int getAbsoluteIndex() {
    return _absoluteIndex;
  }

  public void setAbsoluteIndex(int absoluteIndex) {
    this._absoluteIndex = absoluteIndex;
  }

  public void setIndex(UrlPart urlPart, int index) {
    switch (urlPart) {
      case SCHEME:
        _schemeIndex = index;
        break;
      case USERNAME_PASSWORD:
        _usernamePasswordIndex = index;
        break;
      case HOST:
        _hostIndex = index;
        break;
      case PORT:
        _portIndex = index;
        break;
      case PATH:
        _pathIndex = index;
        break;
      case QUERY:
        _queryIndex = index;
        break;
      case FRAGMENT:
        _fragmentIndex = index;
        break;
      default:
        break;
    }
  }

  /**
   * @param urlPart The part you want the index of
   * @return Returns the index of the part
   */
  public int indexOf(UrlPart urlPart) {
    switch (urlPart) {
      case SCHEME:
        return _schemeIndex;
      case USERNAME_PASSWORD:
        return _usernamePasswordIndex;
      case HOST:
        return _hostIndex;
      case PORT:
        return _portIndex;
      case PATH:
        return _pathIndex;
      case QUERY:
        return _queryIndex;
      case FRAGMENT:
        return _fragmentIndex;
      default:
        return -1;
    }
  }

  public void unsetIndex(UrlPart urlPart) {
    setIndex(urlPart, -1);
  }

  /**
   * This is used in TestUrlMarker to set indices more easily.
   * @param indices array of indices of size 7
   */
  protected UrlMarker setIndices(int[] indices) {
    if (indices == null || indices.length != 7) {
      throw new IllegalArgumentException("Malformed index array.");
    }
    setIndex(UrlPart.SCHEME, indices[0]);
    setIndex(UrlPart.USERNAME_PASSWORD, indices[1]);
    setIndex(UrlPart.HOST, indices[2]);
    setIndex(UrlPart.PORT, indices[3]);
    setIndex(UrlPart.PATH, indices[4]);
    setIndex(UrlPart.QUERY, indices[5]);
    setIndex(UrlPart.FRAGMENT, indices[6]);
    return this;
  }
}


url-detector/src/main/java/com/linkedin/urls/HostNormalizer.java
/**
 * Copyright 2015 LinkedIn Corp. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 */
package com.linkedin.urls;

import com.linkedin.urls.detection.CharUtils;
import java.net.IDN;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.apache.commons.lang3.StringUtils;


/**
 * Normalizes the host by converting hex characters to the actual textual representation, changes ip addresses
 * to a formal format. Then re-encodes the final host name.
 */
public class HostNormalizer {
  private static final long MAX_NUMERIC_DOMAIN_VALUE = 4294967295L;
  private static final int MAX_IPV4_PART = 255;
  private static final int MIN_IP_PART = 0;
  private static final int MAX_IPV6_PART = 0xFFFF;
  private static final int IPV4_MAPPED_IPV6_START_OFFSET = 12;
  private static final int NUMBER_BYTES_IN_IPV4 = 4;

  private byte[] _bytes;
  private String _host;
  private String _normalizedHost;

  public HostNormalizer(String host) {
    _host = host;
    _bytes = null;

    normalizeHost();
  }

  private void normalizeHost() {
    if (StringUtils.isEmpty(_host)) {
      return;
    }

    String host;
    try {
      //replace high unicode characters
      host = IDN.toASCII(_host, IDN.ALLOW_UNASSIGNED);
    } catch (IllegalArgumentException ex) {
      //occurs when the url is invalid. Just return
      return;
    }

    host = host.toLowerCase();
    host = UrlUtil.decode(host);

    _bytes = tryDecodeHostToIp(host);

    if (_bytes != null) {
      InetAddress address;
      try {
        address = InetAddress.getByAddress(_bytes);
        String ipAddress = address.getHostAddress();
        if (address instanceof Inet6Address) {
          host = "[" + ipAddress + "]";
        } else {
          host = ipAddress;
        }
      } catch (UnknownHostException e) {
        return;
      }
    }

    if (StringUtils.isEmpty(host)) {
      return;
    }

    host = UrlUtil.removeExtraDots(host);

    _normalizedHost = UrlUtil.encode(host).replace("\\x", "%");
  }

  /**
   * Checks if the host is an ip address. Returns the byte representation of it
   */
  private byte[] tryDecodeHostToIp(String host) {
    if (host.startsWith("[") && host.endsWith("]")) {
      return tryDecodeHostToIPv6(host);
    }
    return tryDecodeHostToIPv4(host);
  }

  /**
   * This covers cases like:
   * Hexadecimal:        0x1283983
   * Decimal:            12839273
   * Octal:              037362273110
   * Dotted Decimal:     192.168.1.1
   * Dotted Hexadecimal: 0xfe.0x83.0x18.0x1
   * Dotted Octal:       0301.00.046.00
   * Dotted Mixed:       0x38.168.077.1
   *
   * if ipv4 was found, _bytes is set to the byte representation of the ipv4 address
   */
  private byte[] tryDecodeHostToIPv4(String host) {
    String[] parts = CharUtils.splitByDot(host);
    int numParts = parts.length;

    if (numParts != 4 && numParts != 1) {
      return null;
    }

    byte[] bytes = new byte[16];

    //An ipv4 mapped ipv6 bytes will have the 11th and 12th byte as 0xff
    bytes[10] = (byte) 0xff;
    bytes[11] = (byte) 0xff;
    for (int i = 0; i < parts.length; i++) {
      String parsedNum;
      int base;
      if (parts[i].startsWith("0x")) { //hex
        parsedNum = parts[i].substring(2);
        base = 16;
      } else if (parts[i].startsWith("0")) { //octal
        parsedNum = parts[i].substring(1);
        base = 8;
      } else { //decimal
        parsedNum = parts[i];
        base = 10;
      }

      Long section;
      try {
        section = parsedNum.isEmpty() ? 0 : Long.parseLong(parsedNum, base);
      } catch (NumberFormatException e) {
        return null;
      }

      if (numParts == 4 && section > MAX_IPV4_PART || //This would look like 288.1.2.4
          numParts == 1 && section > MAX_NUMERIC_DOMAIN_VALUE || //This would look like 4294967299
          section < MIN_IP_PART) {
        return null;
      }
      //bytes 13->16 is where the ipv4 address of an ipv4-mapped-ipv6-address is stored.
      if (numParts == 4) {
        bytes[IPV4_MAPPED_IPV6_START_OFFSET + i] = section.byteValue();
      } else { //numParts == 1
        int index = IPV4_MAPPED_IPV6_START_OFFSET;
        bytes[index++] = (byte) ((section >> 24) & 0xFF);
        bytes[index++] = (byte) ((section >> 16) & 0xFF);
        bytes[index++] = (byte) ((section >> 8) & 0xFF);
        bytes[index] = (byte) (section & 0xFF);
        return bytes;
      }
    }

    return bytes;
  }

  /**
   * Recommendation for IPv6 Address Text Representation
   * http://tools.ietf.org/html/rfc5952
   *
   * if ipv6 was found, _bytes is set to the byte representation of the ipv6 address
   */
  private byte[] tryDecodeHostToIPv6(String host) {
    String ip = host.substring(1, host.length() - 1);
    List<String> parts = new ArrayList<String>(Arrays.asList(ip.split(":", -1)));
    if (parts.size() < 3) {
      return null;
    }

    //Check for embedded ipv4 address
    String lastPart = parts.get(parts.size() - 1);
    int zoneIndexStart = lastPart.lastIndexOf("%");
    String lastPartWithoutZoneIndex = zoneIndexStart == -1 ? lastPart : lastPart.substring(0, zoneIndexStart);
    byte[] ipv4Address = null;
    if (!isHexSection(lastPartWithoutZoneIndex)) {
      ipv4Address = tryDecodeHostToIPv4(lastPartWithoutZoneIndex);
    }

    byte[] bytes = new byte[16];
    //How many parts do we need to fill by the end of this for loop?
    int totalSize = ipv4Address == null ? 8 : 6;
    //How many zeroes did we fill in the case of double colons? Ex: [::1] will have numberOfFilledZeroes = 7
    int numberOfFilledZeroes = 0;
    //How many sections do we have to parse through? Ex: [fe80:ff::192.168.1.1] size = 3, another ex: [a:a::] size = 4
    int size = ipv4Address == null ? parts.size() : parts.size() - 1;
    for (int i = 0; i < size; i++) {
      int lenPart = parts.get(i).length();
      if (lenPart == 0 && i != 0 && i != parts.size() - 1) {
        numberOfFilledZeroes = totalSize - size;
        for (int k = i; k < numberOfFilledZeroes + i; k++) {
          System.arraycopy(sectionToTwoBytes(0), 0, bytes, k * 2, 2);
        }
      }
      Integer section;
      try {
        section = lenPart == 0 ? 0 : Integer.parseInt(parts.get(i), 16);
      } catch (NumberFormatException e) {
        return null;
      }
      if (section > MAX_IPV6_PART || section < MIN_IP_PART) {
        return null;
      }
      System.arraycopy(sectionToTwoBytes(section), 0, bytes, (numberOfFilledZeroes + i) * 2, 2);
    }

    if (ipv4Address != null) {
      System.arraycopy(ipv4Address, IPV4_MAPPED_IPV6_START_OFFSET, bytes, IPV4_MAPPED_IPV6_START_OFFSET,
          NUMBER_BYTES_IN_IPV4);
    }
    return bytes;
  }

  private static boolean isHexSection(String section) {
    for (int i = 0; i < section.length(); i++) {
      if (!CharUtils.isHex(section.charAt(i))) {
        return false;
      }
    }
    return true;
  }

  private static byte[] sectionToTwoBytes(int section) {
    byte[] bytes = new byte[2];
    bytes[0] = (byte) ((section >> 8) & 0xff);
    bytes[1] = (byte) (section & 0xff);
    return bytes;
  }

  protected byte[] getBytes() {
    return _bytes;
  }

  protected String getNormalizedHost() {
    return _normalizedHost;
  }
}


url-detector/src/main/java/com/linkedin/urls/Url.java
/**
 * Copyright 2015 LinkedIn Corp. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 */
package com.linkedin.urls;

import com.linkedin.urls.detection.UrlDetector;
import com.linkedin.urls.detection.UrlDetectorOptions;
import java.net.MalformedURLException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.commons.lang3.StringUtils;


/**
 * Creating own Uri class since java.net.Uri would throw parsing exceptions
 * for URL's considered ok by browsers.
 *
 * Also to avoid further conflict, this does stuff that the normal Uri object doesn't do:
 * - Converts http://google.com/a/b/.//./../c to http://google.com/a/c
 * - Decodes repeatedly so that http://host/%2525252525252525 becomes http://host/%25 while normal decoders
 *     would make it http://host/%25252525252525 (one less 25)
 * - Removes tabs and new lines: http://www.google.com/foo\tbar\rbaz\n2 becomes "http://www.google.com/foobarbaz2"
 * - Converts IP addresses: http://3279880203/blah becomes http://195.127.0.11/blah
 * - Strips fragments (anything after #)
 *
 */
public class Url {

  private static final String DEFAULT_SCHEME = "http";
  private static final Map<String, Integer> SCHEME_PORT_MAP;
  static {
    SCHEME_PORT_MAP = new HashMap<String, Integer>();
    SCHEME_PORT_MAP.put("http", 80);
    SCHEME_PORT_MAP.put("https", 443);
    SCHEME_PORT_MAP.put("ftp", 21);
  }
  private UrlMarker _urlMarker;
  private String _scheme;
  private String _username;
  private String _password;
  private String _host;
  private int _port = 0;
  private String _path;
  private String _query;
  private String _fragment;
  private String _originalUrl;

  protected Url(UrlMarker urlMarker) {
    _urlMarker = urlMarker;
    _originalUrl = urlMarker.getOriginalUrl();
  }

  /**
   * Returns a url given a single url.
   */
  public static Url create(String url) throws MalformedURLException {
    String formattedString = UrlUtil.removeSpecialSpaces(url.trim().replace(" ", "%20"));
    List<Url> urls = new UrlDetector(formattedString, UrlDetectorOptions.ALLOW_SINGLE_LEVEL_DOMAIN).detect();
    if (urls.size() == 1) {
      return urls.get(0);
    } else if (urls.size() == 0) {
      throw new MalformedURLException("We couldn't find any urls in string: " + url);
    } else {
      throw new MalformedURLException("We found more than one url in string: " + url);
    }
  }

  /**
   * Returns a normalized url given a url object
   */
  public NormalizedUrl normalize() {
    return new NormalizedUrl(_urlMarker);
  }

  @Override
  public String toString() {
    return this.getFullUrl();
  }

  /**
   * Note that this includes the fragment
   * @return Formats the url to: [scheme]://[username]:[password]@[host]:[port]/[path]?[query]#[fragment]
   */
  public String getFullUrl() {
    return getFullUrlWithoutFragment() + StringUtils.defaultString(getFragment());
  }

  /**
   * Get the position of URL in the original string. You can then retrieve URL by substring'ing original
   * text. This can be useful for obtaining context where URL appeared.
   *
   * @return Index of the first character of URL in original string.
   */
  public int getAbsoluteIndex() {
    return _urlMarker.getAbsoluteIndex();
  }

  /**
   *
   * @return Formats the url to: [scheme]://[username]:[password]@[host]:[port]/[path]?[query]
   */
  public String getFullUrlWithoutFragment() {
    StringBuilder url = new StringBuilder();
    if (!StringUtils.isEmpty(getScheme())) {
      url.append(getScheme());
      url.append(":");
    }
    url.append("//");

    if (!StringUtils.isEmpty(getUsername())) {
      url.append(getUsername());
      if (!StringUtils.isEmpty(getPassword())) {
        url.append(":");
        url.append(getPassword());
      }
      url.append("@");
    }

    url.append(getHost());
    if (getPort() > 0 && getPort() != SCHEME_PORT_MAP.get(getScheme())) {
      url.append(":");
      url.append(getPort());
    }

    url.append(getPath());
    url.append(getQuery());

    return url.toString();
  }

  public String getScheme() {
    if (_scheme == null) {
      if (exists(UrlPart.SCHEME)) {
        _scheme = getPart(UrlPart.SCHEME);
        int index = _scheme.indexOf(":");
        if (index != -1) {
          _scheme = _scheme.substring(0, index);
        }
      } else if (!_originalUrl.startsWith("//")) {
        _scheme = DEFAULT_SCHEME;
      }
    }
    return StringUtils.defaultString(_scheme);
  }

  public String getUsername() {
    if (_username == null) {
      populateUsernamePassword();
    }
    return StringUtils.defaultString(_username);
  }

  public String getPassword() {
    if (_password == null) {
      populateUsernamePassword();
    }
    return StringUtils.defaultString(_password);
  }

  public String getHost() {
    if (_host == null) {
      _host = getPart(UrlPart.HOST);
      if (exists(UrlPart.PORT)) {
        _host = _host.substring(0, _host.length() - 1);
      }
    }
    return _host;
  }

  /**
   * port = 0 means it hasn't been set yet. port = -1 means there is no port
   */
  public int getPort() {
    if (_port == 0) {
      String portString = getPart(UrlPart.PORT);
      if (portString != null && !portString.isEmpty()) {
        try {
          _port = Integer.parseInt(portString);
        } catch (NumberFormatException e) {
          _port = -1;
        }
      } else if (SCHEME_PORT_MAP.containsKey(getScheme())) {
        _port = SCHEME_PORT_MAP.get(getScheme());
      } else {
        _port = -1;
      }
    }
    return _port;
  }

  public String getPath() {
    if (_path == null) {
      _path = exists(UrlPart.PATH) ? getPart(UrlPart.PATH) : "/";
    }
    return _path;
  }

  public String getQuery() {
    if (_query == null) {
      _query = getPart(UrlPart.QUERY);
    }
    return StringUtils.defaultString(_query);
  }

  public String getFragment() {
    if (_fragment == null) {
      _fragment = getPart(UrlPart.FRAGMENT);
    }
    return StringUtils.defaultString(_fragment);
  }

  /**
   * Always returns null for non normalized urls.
   */
  public byte[] getHostBytes() {
    return null;
  }

  public String getOriginalUrl() {
    return _originalUrl;
  }

  private void populateUsernamePassword() {
    if (exists(UrlPart.USERNAME_PASSWORD)) {
      String usernamePassword = getPart(UrlPart.USERNAME_PASSWORD);
      String[] usernamePasswordParts = usernamePassword.substring(0, usernamePassword.length() - 1).split(":");
      if (usernamePasswordParts.length == 1) {
        _username = usernamePasswordParts[0];
      } else if (usernamePasswordParts.length == 2) {
        _username = usernamePasswordParts[0];
        _password = usernamePasswordParts[1];
      }
    }
  }

  /**
   * @param urlPart The url part we are checking for existence
   * @return Returns true if the part exists.
   */
  private boolean exists(UrlPart urlPart) {
    return urlPart != null && _urlMarker.indexOf(urlPart) >= 0;
  }

  /**
   * For example, in http://yahoo.com/lala/, nextExistingPart(UrlPart.HOST) would return UrlPart.PATH
   * @param urlPart The current url part
   * @return Returns the next part; if there is no existing next part, it returns null
   */
  private UrlPart nextExistingPart(UrlPart urlPart) {
    UrlPart nextPart = urlPart.getNextPart();
    if (exists(nextPart)) {
      return nextPart;
    } else if (nextPart == null) {
      return null;
    } else {
      return nextExistingPart(nextPart);
    }
  }

  /**
   * @param part The part that we want. Ex: host, path
   */
  private String getPart(UrlPart part) {
    if (!exists(part)) {
      return null;
    }

    UrlPart nextPart = nextExistingPart(part);
    if (nextPart == null) {
      return _originalUrl.substring(_urlMarker.indexOf(part));
    }
    return _originalUrl.substring(_urlMarker.indexOf(part), _urlMarker.indexOf(nextPart));
  }

  protected void setRawPath(String path) {
    _path = path;
  }

  protected void setRawHost(String host) {
    _host = host;
  }

  protected String getRawPath() {
    return _path;
  }

  protected String getRawHost() {
    return _host;
  }

  protected UrlMarker getUrlMarker() {
    return _urlMarker;
  }
}
